akd_audit_quorum/src/crypto/mod.rs

```diff
@@ -0,0 +1,309 @@
+    async fn generate_encrypted_shards(
+        &self,
+        shards: Vec<EncryptedQuorumKeyShard>,
+        node_public_keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<EncryptedQuorumKeyShard>, QuorumOperationError>;
+
+    /// Encrypt the given material using the provided public key, optionally with the provided nonce
+    async fn encrypt_message(
+        &self,
+        public_key: Vec<u8>,
+        plaintext: Vec<u8>,
+        nonce: Nonce,
+    ) -> Result<Vec<u8>, QuorumOperationError>;
+
+    /// Decrypt the specified message utilizing this node's
+    /// private key. This will require that the message contain
+    /// a nonce. I.e. this cannot be used to decrypt the encrypted
+    /// shard partials
+    async fn decrypt_message(
+        &self,
+        cyphertext: Vec<u8>,
+    ) -> Result<(Vec<u8>, Nonce), QuorumOperationError>;
+
+    /// Generate a commitment on the epoch changes using the quorum key
+    async fn generate_commitment<H: Hasher>(
+        &self,
+        quorum_key_shards: Vec<EncryptedQuorumKeyShard>,
+        epoch: u64,
+        previous_hash: H::Digest,
+        current_hash: H::Digest,
+    ) -> Result<QuorumCommitment<H>, QuorumOperationError>;
+
+    /// Validate the commitment applied on the specified epoch settings
+    async fn validate_commitment<H: Hasher>(
+        public_key: Vec<u8>,
+        commitment: QuorumCommitment<H>,
+    ) -> Result<bool, QuorumOperationError>;
+
+    // ==================================================================
+    // Common trait logic
+    // ==================================================================
+
+    /// Get the number of shards required to reconstruct the quorum key
+    fn shards_required(n: u8) -> u8 {
+        let f = if (n - 1) % 3 == 0 {
+            (n - 1) / 3
+        } else {
+            // there's remainders
+            1 + (n - 1) / 3
+        };
+        2 * f + 1
+    }
+
+    /// Generate num_shards shards of the quorum key, and return the shard pieces.
+    /// We take ownership of the quorum key here to help prevent leakage of the key.
+    /// By taking ownership, someone needs to explicitely clone it to use it elsewhere
+    fn generate_shards(
+        quorum_key: [u8; QUORUM_KEY_SIZE],
+        num_shards: u8,
+    ) -> Result<Vec<QuorumKeyShard>, QuorumOperationError> {
+        let num_approvals = Self::shards_required(num_shards);
+
+        let mut parts = vec![vec![]; num_shards.into()];
+
+        for i in 0..QUORUM_KEY_NUM_PARTS {
+            let part: [u8; DATA_SIZE] = quorum_key[i * DATA_SIZE..(i + 1) * DATA_SIZE]
+                .try_into()
+                .map_err(|_| {
+                QuorumOperationError::Sharding(format!(
+                    "Unable to convert quorum key slice into SSS shardable component of len {}",
+                    DATA_SIZE
+                ))
+            })?;
+            let results = create_shares(&part, num_shards, num_approvals)?;
```

Comment:
How about to add immediate verification of shares?

let restored = combine_shares(&results).unwrap();
assert_eq!(restored, Some(part));

Reply:
Maybe something along those lines, but we generally stay away from unwrap and asserts outside of test env's as they can cause a ```panic!``` in prod code, which will be unhandled.

Reply:
I'd better validate key could be restored and panic! if not rather than distribute corrupted parts.